python/pyspark/sql/session.py

```diff
@@ -225,6 +225,7 @@ def __init__(self, sparkContext, jsparkSession=None):
         if SparkSession._instantiatedSession is None \
                 or SparkSession._instantiatedSession._sc._jsc is None:
             SparkSession._instantiatedSession = self
+            self._jvm.org.apache.spark.sql.SparkSession.setDefaultSession(self._jsparkSession)
```

Comment:
I know it looks a bit odd test but should we better have a test like

```
spark._jvm.org.apache.spark.sql.SparkSession.getDefaultSession().nonEmpty()
```

?

Reply:
`setActiveSession` or `setDefaultSession`? Which one is more proper to set here?

Btw, shall we clear it when stopping PySpark SparkSession?

Reply:
By looking at Scala code, seems Scala `getOrCreate` will set to `defaultSession` I was thinking it is more proper to set to `defaultSession`. Does PySpark support multiple sessions?

Reply:
>Btw, shall we clear it when stopping PySpark SparkSession?

JVM SparkSession will clear it when application is stopped (https://github.com/apache/spark/blob/3e252514741447004f3c18ddd77c617b4e37cfaa/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L961).

Reply:
Actually, It seems not because we don't call this code path. Stop and start logic is convoluted in PySpark in my humble opinion. Setting the default one fixes an actual issue and seems we are okay with it, at least.

Reply:
Ohh, I see. My misunderstanding. let me figure out a way to clear this object.

Reply:
Simplest way I can think of is just add it in `def stop`:

```
self._jvm.org.apache.spark.sql.SparkSession.clearDefaultSession()
```

Reply:
Yeah, I think we can clear it in `def stop`.

Reply:
I think this is somewhat conflicted ... If I can SparkSession.getOrCreate in python and JVM already has a defaultSession, what should happen?
I suppose one approach would be to keep python independent.. but with this change it would overwrite the defaultSession that might be valid?

Reply:
I think it would overwrite. Can we try something like this?

```diff
diff --git a/python/pyspark/sql/session.py b/python/pyspark/sql/session.py
index 6c84023c43f..0bdfc88153f 100644
--- a/python/pyspark/sql/session.py
+++ b/python/pyspark/sql/session.py
@@ -213,7 +213,10 @@ class SparkSession(object):
         self._jsc = self._sc._jsc
         self._jvm = self._sc._jvm
         if jsparkSession is None:
-            jsparkSession = self._jvm.SparkSession(self._jsc.sc())
+            if self._jvm.SparkSession.getDefaultSession().isEmpty():
+                jsparkSession = self._jvm.SparkSession.getDefaultSession().get()
+            else:
+                jsparkSession = self._jvm.SparkSession(self._jsc.sc())
         self._jsparkSession = jsparkSession
         self._jwrapped = self._jsparkSession.sqlContext()
         self._wrapped = SQLContext(self._sc, self, self._jwrapped)
@@ -225,6 +228,8 @@ class SparkSession(object):
         if SparkSession._instantiatedSession is None \
                 or SparkSession._instantiatedSession._sc._jsc is None:
             SparkSession._instantiatedSession = self
+            if self._jvm.SparkSession.getDefaultSession().isEmpty():
+               self._jvm.SparkSession.setDefaultSession(self._jsparkSession)

     def _repr_html_(self):
         return """
@@ -759,6 +764,7 @@ class SparkSession(object):
         """Stop the underlying :class:`SparkContext`.
         """
         self._sc.stop()
+        self._jvm.org.apache.spark.sql.SparkSession.clearDefaultSession()
         SparkSession._instantiatedSession = None
```

Reply:
Thanks @HyukjinKwon , let me update the code.